Sources/CodeEditSourceEditor/Gutter/LineFolding/LineFoldingModel.swift

```diff
@@ -0,0 +1,155 @@
+//
+//  LineFoldingModel.swift
+//  CodeEditSourceEditor
+//
+//  Created by Khan Winter on 5/7/25.
+//
+
+import AppKit
+import CodeEditTextView
+
+/// # Basic Premise
+///
+/// We need to update, delete, or add fold ranges in the invalidated lines.
+///
+/// # Implementation
+///
+/// - For each line in the document, put its indent level into a list.
+/// - Loop through the list, creating nested folds as indents go up and down.
+///
+class LineFoldingModel: NSObject, NSTextStorageDelegate {
+    /// An ordered tree of fold ranges in a document. Can be traversed using ``FoldRange/parent``
+    /// and ``FoldRange/subFolds``.
+    private var foldCache: [FoldRange] = []
+
+    weak var foldProvider: LineFoldProvider?
+    weak var textView: TextView?
+
+    init(textView: TextView, foldProvider: LineFoldProvider?) {
+        self.textView = textView
+        self.foldProvider = foldProvider
+        super.init()
+        textView.addStorageDelegate(self)
+        buildFoldsForDocument()
+    }
+
+    func getFolds(in lineRange: ClosedRange<Int>) -> [FoldRange] {
+        foldCache.filter({ $0.lineRange.overlaps(lineRange) })
+    }
+
+    /// Build out the ``foldCache`` for the entire document.
+    ///
+    /// For each line in the document, find the indentation level using the ``levelProvider``. At each line, if the
+    /// indent increases from the previous line, we start a new fold. If it decreases we end the fold we were in.
```

Comment:
We may look into other methods of determining ranges in addition to indentation. For example, multiline comments might need to be folded which will exist at the same indentation level.

Reply:
Agreed, that's why I structured it to have a 'provider' type that should be able to provide folding ranges using something like tree-sitter or regex.

I may scrap this recursive data structure too. I'm torn because it's very very nice to have for the UI drawing code, but is proving hard to update live with edits. That can be changed later, however.